File: cli/python/commands/infra.py
"""
nexus infra — Infrastructure lifecycle management.

Commands:
  start      Start all services (docker compose up)
  stop       Stop all services
  restart    Restart one or all services
  scale      Scale a service to N replicas
  status     Show running container status
  logs       Stream/tail logs for a service
  cleanup    Stop and remove containers + volumes
  k8s        Kubernetes subgroup (apply, delete, status, scale)
  build      Build Docker images
"""

import subprocess
import sys
import click
from rich.console import Console
from rich.table import Table
from rich import box

console = Console()

COMPOSE_FILE     = "docker-compose.yml"
COMPOSE_DEV_FILE = "docker-compose.dev.yml"
SERVICES = ["web", "python-service", "ai-service", "postgres", "redis", "mongo", "nginx"]


def _run(cmd: list[str], check=True, stream=False):
    """Run a subprocess, optionally streaming stdout."""
    console.print(f"[dim]$ {' '.join(cmd)}[/]")
    if stream:
        proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=sys.stderr)
        proc.wait()
        if check and proc.returncode != 0:
            raise SystemExit(proc.returncode)
    else:
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.stdout:
            console.print(result.stdout)
        if result.stderr:
            console.print(f"[dim red]{result.stderr}[/]")
        if check and result.returncode != 0:
            raise SystemExit(result.returncode)
        return result


@click.group()
def infra():
    """Docker Compose and Kubernetes infrastructure lifecycle management."""
    pass


# ── start ─────────────────────────────────────────────────────────────────────

@infra.command()
@click.option("--dev",     is_flag=True, help="Use dev compose override (hot-reload).")
@click.option("--detach",  "-d", is_flag=True, default=True, show_default=True, help="Run in background.")
@click.option("--service", "-s", type=click.Choice(SERVICES), default=None, help="Start a specific service only.")
@click.option("--build",   "-b", is_flag=True, help="Rebuild images before starting.")
@click.help_option("--help", "-h")
def start(dev, detach, service, build):
    """
    Start all NexusConsult services via Docker Compose.

    \b
    Examples:
      nexus infra start
      nexus infra start --dev
      nexus infra start --build
      nexus infra start --service web
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE]
    if dev:
        cmd += ["-f", COMPOSE_DEV_FILE]
    cmd += ["up"]
    if detach:
        cmd += ["-d"]
    if build:
        cmd += ["--build"]
    if service:
        cmd.append(service)
    console.print(f"[cyan]Starting services{'(dev)' if dev else ''}…[/]")
    _run(cmd, stream=True)
    console.print("[green]✓ Services started.[/]")


# ── stop ──────────────────────────────────────────────────────────────────────

@infra.command()
@click.option("--service", "-s", type=click.Choice(SERVICES), default=None, help="Stop a specific service only.")
@click.help_option("--help", "-h")
def stop(service):
    """
    Stop all (or a specific) running services.

    \b
    Examples:
      nexus infra stop
      nexus infra stop --service ai-service
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "stop"]
    if service:
        cmd.append(service)
    _run(cmd, stream=True)
    console.print("[green]✓ Services stopped.[/]")


# ── restart ───────────────────────────────────────────────────────────────────

@infra.command()
@click.option("--service", "-s", type=click.Choice(SERVICES), default=None, help="Restart a specific service.")
@click.help_option("--help", "-h")
def restart(service):
    """
    Restart all (or a specific) services.

    \b
    Examples:
      nexus infra restart
      nexus infra restart --service web
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "restart"]
    if service:
        cmd.append(service)
    _run(cmd, stream=True)
    console.print("[green]✓ Services restarted.[/]")


# ── scale ─────────────────────────────────────────────────────────────────────

@infra.command()
@click.argument("service", type=click.Choice(SERVICES))
@click.argument("replicas", type=int)
@click.help_option("--help", "-h")
def scale(service, replicas):
    """
    Scale SERVICE to REPLICAS instances.

    \b
    Examples:
      nexus infra scale web 3
      nexus infra scale python-service 5
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "scale", f"{service}={replicas}"]
    _run(cmd, stream=True)
    console.print(f"[green]✓ {service} scaled to {replicas} replica(s).[/]")


# ── status ────────────────────────────────────────────────────────────────────

@infra.command()
@click.help_option("--help", "-h")
def status():
    """
    Show status of all running containers.

    \b
    Examples:
      nexus infra status
    """
    _run(["docker", "compose", "-f", COMPOSE_FILE, "ps"], stream=True)


# ── logs ──────────────────────────────────────────────────────────────────────

@infra.command()
@click.argument("service", type=click.Choice(SERVICES))
@click.option("--tail", "-n", type=int, default=100, show_default=True, help="Number of lines to show.")
@click.option("--follow", "-f", is_flag=True, help="Follow log output (streaming).")
@click.help_option("--help", "-h")
def logs(service, tail, follow):
    """
    View logs for a specific SERVICE.

    \b
    Examples:
      nexus infra logs web --tail 50
      nexus infra logs ai-service --follow
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "logs", f"--tail={tail}"]
    if follow:
        cmd.append("-f")
    cmd.append(service)
    _run(cmd, stream=True)


# ── cleanup ───────────────────────────────────────────────────────────────────

@infra.command()
@click.option("--volumes", "-v", is_flag=True, help="Also remove named volumes (destroys data).")
@click.option("--yes", "-y", is_flag=True,     help="Skip confirmation prompt.")
@click.help_option("--help", "-h")
def cleanup(volumes, yes):
    """
    Stop and remove all containers, networks, and optionally volumes.

    \b
    Examples:
      nexus infra cleanup
      nexus infra cleanup --volumes --yes
    """
    if not yes:
        msg = "Remove all containers" + (" AND volumes (data loss!)" if volumes else "") + "?"
        click.confirm(msg, abort=True)
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "down"]
    if volumes:
        cmd.append("-v")
    _run(cmd, stream=True)
    console.print("[green]✓ Cleanup complete.[/]")


# ── build ─────────────────────────────────────────────────────────────────────

@infra.command()
@click.option("--service", "-s", type=click.Choice(["web", "python-service", "ai-service"]),
              default=None, help="Build a specific service image.")
@click.option("--no-cache", is_flag=True, help="Disable Docker layer cache.")
@click.help_option("--help", "-h")
def build(service, no_cache):
    """
    Build Docker images for the project.

    \b
    Examples:
      nexus infra build
      nexus infra build --service ai-service --no-cache
    """
    cmd = ["docker", "compose", "-f", COMPOSE_FILE, "build"]
    if no_cache:
        cmd.append("--no-cache")
    if service:
        cmd.append(service)
    _run(cmd, stream=True)
    console.print("[green]✓ Build complete.[/]")


# ── k8s subgroup ─────────────────────────────────────────────────────────────

@infra.group()
def k8s():
    """Kubernetes cluster management (apply, delete, status, scale, logs)."""
    pass


@k8s.command()
@click.option("--dir", "manifest_dir", default="k8s", show_default=True, help="K8s manifests directory.")
@click.help_option("--help", "-h")
def apply(manifest_dir):
    """
    Apply all Kubernetes manifests to the cluster.

    \b
    Examples:
      nexus infra k8s apply
      nexus infra k8s apply --dir k8s/staging
    """
    _run(["kubectl", "apply", "-f", manifest_dir], stream=True)


@k8s.command()
@click.option("--dir", "manifest_dir", default="k8s", show_default=True, help="K8s manifests directory.")
@click.option("--yes", "-y", is_flag=True, help="Skip confirmation.")
@click.help_option("--help", "-h")
def delete(manifest_dir, yes):
    """
    Delete Kubernetes resources defined in manifests.

    \b
    Examples:
      nexus infra k8s delete
    """
    if not yes:
        click.confirm("Delete all K8s resources?", abort=True)
    _run(["kubectl", "delete", "-f", manifest_dir], stream=True)


@k8s.command("status")
@click.option("--namespace", "-n", default="nexus", show_default=True)
@click.help_option("--help", "-h")
def k8s_status(namespace):
    """
    Show pod status in the nexus namespace.

    \b
    Examples:
      nexus infra k8s status
    """
    _run(["kubectl", "get", "pods", "-n", namespace, "-o", "wide"], stream=True)


@k8s.command("scale")
@click.argument("deployment")
@click.argument("replicas", type=int)
@click.option("--namespace", "-n", default="nexus", show_default=True)
@click.help_option("--help", "-h")
def k8s_scale(deployment, replicas, namespace):
    """
    Scale a Kubernetes DEPLOYMENT to REPLICAS.

    \b
    Examples:
      nexus infra k8s scale web-deployment 5
      nexus infra k8s scale python-deployment 3 -n nexus
    """
    _run(["kubectl", "scale", "deployment", deployment,
          f"--replicas={replicas}", "-n", namespace], stream=True)

File: cli/go/cmd/infra.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/spf13/cobra"
)

const composeFile = "docker-compose.yml"

var infraCmd = &cobra.Command{
	Use:   "infra",
	Short: "Docker Compose and Kubernetes infrastructure lifecycle management",
}

func runCmd(args ...string) error {
	fmt.Printf("%s $ %s\n", dim("→"), strings.Join(args, " "))
	cmd := exec.Command(args[0], args[1:]...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

// ── start ─────────────────────────────────────────────────────────────────────

var (
	startDev     bool
	startBuild   bool
	startService string
)

var infraStartCmd = &cobra.Command{
	Use:   "start",
	Short: "Start all Docker Compose services",
	Long: `Start NexusConsult services via Docker Compose.

Examples:
  nexus infra start
  nexus infra start --dev
  nexus infra start --build
  nexus infra start --service web`,
	RunE: func(cmd *cobra.Command, args []string) error {
		c := []string{"docker", "compose", "-f", composeFile}
		if startDev {
			c = append(c, "-f", "docker-compose.dev.yml")
		}
		c = append(c, "up", "-d")
		if startBuild { c = append(c, "--build") }
		if startService != "" { c = append(c, startService) }
		return runCmd(c...)
	},
}

// ── stop ──────────────────────────────────────────────────────────────────────

var stopService string
var infraStopCmd = &cobra.Command{
	Use:   "stop",
	Short: "Stop running services",
	Long:  "Examples:\n  nexus infra stop\n  nexus infra stop --service ai-service",
	RunE: func(cmd *cobra.Command, args []string) error {
		c := []string{"docker", "compose", "-f", composeFile, "stop"}
		if stopService != "" { c = append(c, stopService) }
		return runCmd(c...)
	},
}

// ── restart ───────────────────────────────────────────────────────────────────

var restartService string
var infraRestartCmd = &cobra.Command{
	Use:   "restart",
	Short: "Restart services",
	Long:  "Examples:\n  nexus infra restart\n  nexus infra restart --service web",
	RunE: func(cmd *cobra.Command, args []string) error {
		c := []string{"docker", "compose", "-f", composeFile, "restart"}
		if restartService != "" { c = append(c, restartService) }
		return runCmd(c...)
	},
}

// ── scale ─────────────────────────────────────────────────────────────────────

var infraScaleCmd = &cobra.Command{
	Use:   "scale <service> <replicas>",
	Short: "Scale a service to N replicas",
	Long:  "Examples:\n  nexus infra scale web 3\n  nexus infra scale python-service 5",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("docker", "compose", "-f", composeFile, "scale", args[0]+"="+args[1])
	},
}

// ── status ────────────────────────────────────────────────────────────────────

var infraStatusCmd = &cobra.Command{
	Use:   "status",
	Short: "Show container status",
	Long:  "Example:\n  nexus infra status",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("docker", "compose", "-f", composeFile, "ps")
	},
}

// ── logs ──────────────────────────────────────────────────────────────────────

var (
	logsTail   int
	logsFollow bool
)

var infraLogsCmd = &cobra.Command{
	Use:   "logs <service>",
	Short: "View logs for a service",
	Long:  "Examples:\n  nexus infra logs web --tail 50\n  nexus infra logs ai-service --follow",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		c := []string{"docker", "compose", "-f", composeFile, "logs", fmt.Sprintf("--tail=%d", logsTail)}
		if logsFollow { c = append(c, "-f") }
		c = append(c, args[0])
		return runCmd(c...)
	},
}

// ── cleanup ───────────────────────────────────────────────────────────────────

var (
	cleanupVolumes bool
	cleanupYes     bool
)

var infraCleanupCmd = &cobra.Command{
	Use:   "cleanup",
	Short: "Stop and remove containers, networks, optionally volumes",
	Long:  "Examples:\n  nexus infra cleanup\n  nexus infra cleanup --volumes --yes",
	RunE: func(cmd *cobra.Command, args []string) error {
		if !cleanupYes {
			msg := "Remove all containers?"
			if cleanupVolumes { msg = "Remove all containers AND volumes (data loss!)?" }
			fmt.Printf("%s [y/N]: ", msg)
			var c string
			fmt.Scan(&c)
			if strings.ToLower(c) != "y" { fmt.Println("Aborted."); return nil }
		}
		c := []string{"docker", "compose", "-f", composeFile, "down"}
		if cleanupVolumes { c = append(c, "-v") }
		return runCmd(c...)
	},
}

// ── build ─────────────────────────────────────────────────────────────────────

var (
	buildService string
	buildNoCache bool
)

var infraBuildCmd = &cobra.Command{
	Use:   "build",
	Short: "Build Docker images",
	Long:  "Examples:\n  nexus infra build\n  nexus infra build --service ai-service --no-cache",
	RunE: func(cmd *cobra.Command, args []string) error {
		c := []string{"docker", "compose", "-f", composeFile, "build"}
		if buildNoCache { c = append(c, "--no-cache") }
		if buildService != "" { c = append(c, buildService) }
		return runCmd(c...)
	},
}

// ── k8s ───────────────────────────────────────────────────────────────────────

var k8sCmd = &cobra.Command{
	Use:   "k8s",
	Short: "Kubernetes cluster management",
}

var k8sApplyCmd = &cobra.Command{
	Use:  "apply",
	Short: "Apply K8s manifests",
	Long:  "Example:\n  nexus infra k8s apply",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("kubectl", "apply", "-f", "k8s")
	},
}

var k8sDeleteCmd = &cobra.Command{
	Use:   "delete",
	Short: "Delete K8s resources",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("kubectl", "delete", "-f", "k8s")
	},
}

var k8sStatusCmd = &cobra.Command{
	Use:   "status",
	Short: "Show K8s pod status",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("kubectl", "get", "pods", "-n", "nexus", "-o", "wide")
	},
}

var k8sScaleCmd = &cobra.Command{
	Use:   "scale <deployment> <replicas>",
	Short: "Scale a K8s deployment",
	Long:  "Example:\n  nexus infra k8s scale web-deployment 5",
	Args:  cobra.ExactArgs(2),
	RunE: func(cmd *cobra.Command, args []string) error {
		return runCmd("kubectl", "scale", "deployment", args[0], "--replicas="+args[1], "-n", "nexus")
	},
}

func init() {
	infraStartCmd.Flags().BoolVar(&startDev,         "dev",     false,  "Use dev compose override")
	infraStartCmd.Flags().BoolVarP(&startBuild,      "build",   "b",    false, "Rebuild images before start")
	infraStartCmd.Flags().StringVarP(&startService,  "service", "s",    "",    "Start specific service only")
	infraStopCmd.Flags().StringVarP(&stopService,    "service", "s",    "",    "Stop specific service only")
	infraRestartCmd.Flags().StringVarP(&restartService, "service", "s", "",    "Restart specific service")
	infraLogsCmd.Flags().IntVarP(&logsTail,          "tail",    "n",    100,   "Number of lines to show")
	infraLogsCmd.Flags().BoolVarP(&logsFollow,       "follow",  "f",    false, "Follow log output")
	infraCleanupCmd.Flags().BoolVarP(&cleanupVolumes,"volumes", "v",    false, "Also remove volumes")
	infraCleanupCmd.Flags().BoolVarP(&cleanupYes,    "yes",     "y",    false, "Skip confirmation")
	infraBuildCmd.Flags().StringVarP(&buildService,  "service", "s",    "",    "Build specific service")
	infraBuildCmd.Flags().BoolVar(&buildNoCache,     "no-cache",        false, "Disable build cache")

	k8sCmd.AddCommand(k8sApplyCmd, k8sDeleteCmd, k8sStatusCmd, k8sScaleCmd)
	infraCmd.AddCommand(infraStartCmd, infraStopCmd, infraRestartCmd, infraScaleCmd,
		infraStatusCmd, infraLogsCmd, infraCleanupCmd, infraBuildCmd, k8sCmd)
}
